### PR TITLE
[MemCpyOpt] Fix the invalid code modification for GEP

### PR DESCRIPTION
**Pull request by kaiyan96**
Relocate the GEP modification to a later stage of the function performCallSlotOption, ensuring that the code remains unchanged if the optimization fails.

### CODE REVIEW

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  %7 = getelementptr inbounds %"struct.MaskedType", ptr %2, i32 0, i32 1
+  store i32 %6, ptr %7, align 4
+  %8 = load i32, ptr %1, align 4
+  %cond = icmp eq i32 %8, 0
+  br i1 %cond, label %4, label %9
+9:                                                ; preds = %4
+  ret i32 0
+}
+
+attributes #0 = { nounwind }
```
Comment (dtcxzyw) [inline llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll:39]: Please add a new line at EOF and reduce the test with `llvm-reduce`. It is obvious that `@.str` is unnecessary.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll [hunk 2]
```diff
@@ -0,0 +1,39 @@
+; RUN: opt -S -passes=memcpyopt < %s -verify-memoryssa | FileCheck %s
```
Comment (nikic) [inline llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll:1]: Use update_test_checks.py.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+%struct.MaskedType = type { i32, i32 }
+
+@.str = private unnamed_addr constant [4 x i8] c"%d\0A\00", align 1
+
+declare void @llvm.lifetime.start.p0(i64, i8* nocapture) #0
+declare void @llvm.lifetime.end.p0(i64, i8* nocapture) #0
+declare void @MaskedFunction(%struct.MaskedType*, i32 addrspace(1)*)
+
+define i32 @test_gep_modified(ptr noundef %0, ptr noundef %1) {
+; CHECK-NOT: LLVM ERROR: Function @test_gep_modified changed by MemCpyOptPass without invalidating analyses
```
Comment (nikic) [inline llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll:15]: This is checked implicitly.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+entry:
+  %2 = alloca %struct.MaskedType, align 4
+  %3 = alloca i32, align 4
+  br label %4
+
+4:                                                ; preds = %4, %entry
+  call void @llvm.lifetime.start.p0(i64 4, ptr %3) #0
+  %5 = addrspacecast i32* %3 to i32 addrspace(1)*
+  call void @MaskedFunction(%struct.MaskedType* noundef %1, i32 addrspace(1)* noundef %5)
```
Comment (nikic) [inline llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll:27]: Use opaque pointers.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+define i32 @test_gep_modified(ptr noundef %0, ptr noundef %1) {
+; CHECK-NOT: LLVM ERROR: Function @test_gep_modified changed by MemCpyOptPass without invalidating analyses
+; CHECK:   call void @MaskedFunction
+; CHECK:   %7 = getelementptr inbounds %struct.MaskedType
+
+entry:
+  %2 = alloca %struct.MaskedType, align 4
+  %3 = alloca i32, align 4
+  br label %4
```
Comment (nikic) [inline llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll:22]: Name instructions and blocks in test.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  call void @llvm.lifetime.start.p0(i64 4, ptr %3) #0
+  %5 = addrspacecast i32* %3 to i32 addrspace(1)*
+  call void @MaskedFunction(%struct.MaskedType* noundef %1, i32 addrspace(1)* noundef %5)
+  %6 = load i32, ptr %3, align 4
+  call void @llvm.lifetime.end.p0(i64 4, ptr %3) #0
+  %7 = getelementptr inbounds %"struct.MaskedType", ptr %2, i32 0, i32 1
+  store i32 %6, ptr %7, align 4
+  %8 = load i32, ptr %1, align 4
+  %cond = icmp eq i32 %8, 0
+  br i1 %cond, label %4, label %9
```
Comment (nikic) [inline llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll:34]: This control flow does not seem relevant to what you're testing.

**Review by nikic — CHANGES_REQUESTED** [review]
Instead of rolling back, we should not do the change until we know the transform will succeed. The simplest way to do that would be to move this check to the end. (Setting a variable to perform the move later would also work.)

File: llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp [hunk 7]
```diff
@@ -1062,19 +1066,25 @@ bool MemCpyOptPass::performCallSlotOptzn(Instruction *cpyLoad,
   // If necessary, perform additional analysis.
   if (isModOrRefSet(MR))
     MR = BAA.callCapturesBefore(C, DestWithSrcSize, DT);
-  if (isModOrRefSet(MR))
+  if (isModOrRefSet(MR)) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:1069]: Please add tests for these three cases which revert the code motion. I'm not sure an RAII guard is needed here. It will revert the change by default unless we explicitly keep the change in the end.

**Comment by kaiyan96** [comment]
> Instead of rolling back, we should not do the change until we know the transform will succeed. The simplest way to do that would be to move this check to the end. (Setting a variable to perform the move later would also work.)

According to this modification(https://reviews.llvm.org/D89623), which claim that "When performing a call slot optimization to a GEP dest, it will currently usually not apply, because the GEP is directly before the memcpy and as such does not dominate the call. We should move it above the call if that satisfies the domination requirement." I thought it is better to apply this move before some further checks.

File: llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp (hunk 7 above)
Comment (kaiyan96) [reply llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp]: RAII is a good idea. I have changed my implementation to an InstructionPositionGuard.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 2 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: Thanks for reply. I used it in my new modification.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 3 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: Fixed

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 4 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: Fixed

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 5 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: Fixed

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 6 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: Fixed

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 1 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: Fixed

**Comment by nikic** [comment]
> > Instead of rolling back, we should not do the change until we know the transform will succeed. The simplest way to do that would be to move this check to the end. (Setting a variable to perform the move later would also work.)
> 
> @nikic According to this modification(https://reviews.llvm.org/D89623), which claim that "When performing a call slot optimization to a GEP dest, it will currently usually not apply, because the GEP is directly before the memcpy and as such does not dominate the call. We should move it above the call if that satisfies the domination requirement." It seems that it is better to apply this move before some further checks.

The further checks do not depend on the instruction being moved. It just needs to be moved at some point if the transform is applied, it does not have to be at that specific place in the sequence of checks.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 2 above)
Comment (nikic) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: It doesn't look that way. Please use update_test_checks.py to generate the CHECK lines and then **do not manually edit them**.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 4 above)
Comment (nikic) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: I still see a lot of non-opaque pointers like `i8*` in here. All of these should be replaced with `ptr`.

File: llvm/test/Transforms/MemCpyOpt/memcpy-gep-position-guard.ll [hunk 8]
```diff
@@ -0,0 +1,51 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
Comment (dtcxzyw) [inline llvm/test/Transforms/MemCpyOpt/memcpy-gep-position-guard.ll:1]: I cannot reproduce the crash `LLVM ERROR: Function @test_gep_modified changed by MemCpyOptPass without invalidating analyses` on the trunk build.

Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy-gep-position-guard.ll:1]: I add '-DLLVM_ENABLE_EXPENSIVE_CHECKS=true' to my cmake, this might cause the ERROR.

File: llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll (hunk 4 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy_invalid_modify.ll]: I apologize for my misunderstanding. I have fixed it now.

File: llvm/test/Transforms/MemCpyOpt/memcpy-gep-modification.ll [hunk 9]
```diff
@@ -0,0 +1,63 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt -S -passes=memcpyopt < %s -verify-memoryssa | FileCheck %s
+
+; ModuleID = 'memcpy_invalid_modify.ll'
+source_filename = "memcpy_invalid_modify.ll"
```
Comment (nikic) [inline llvm/test/Transforms/MemCpyOpt/memcpy-gep-modification.ll:5]: Drop ModuleID/source_filename

File: llvm/test/Transforms/MemCpyOpt/memcpy-gep-modification.ll [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+%struct.MaskedType = type { i8, i8 }
+
+declare void @llvm.lifetime.start.p0(i64, ptr nocapture) #0
+declare void @llvm.lifetime.end.p0(i64, ptr nocapture) #0
+declare void @MaskedFunction1(ptr, ptr addrspace(1))
+declare void @MaskedFunction2(ptr, ptr)
+
+define i8 @test_gep_not_modified(ptr noundef %in0, ptr noundef %in1) {
```
Comment (nikic) [inline llvm/test/Transforms/MemCpyOpt/memcpy-gep-modification.ll:14]: Drop noundef here and elsewhere.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Transforms/MemCpyOpt/memcpy-gep-modification.ll (hunk 10 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy-gep-modification.ll]: Fixed

File: llvm/test/Transforms/MemCpyOpt/memcpy-gep-modification.ll (hunk 9 above)
Comment (kaiyan96) [reply llvm/test/Transforms/MemCpyOpt/memcpy-gep-modification.ll]: Fixed

**Comment by lcvon007** [comment]
> Instead of rolling back, we should not do the change until we know the transform will succeed. The simplest way to do that would be to move this check to the end. (Setting a variable to perform the move later would also work.)

please update commit message too.
